File: searching_algos/BFS.js
// === BFS: Breadth First Search ===

// Strategy: Start with root node, then look left to right at the second level, then left to right on the third level (so on and so forth) until the target value is found. 

// Big O:
    // Time: O(n)

// IMPLEMENTATION:

const breadthFirstSearch = function () {
    let currentNode = this.root;
    let list = [];
    let queue = [];
    queue.push(currentNode);

    while (queue.length > 0) {
        currentNode = queue.shift();
        list.push(currentNode.value);
        if(currentNode.left) {
            queue.push(currentNode.left);
        }
        if(currentNode.right) {
            queue.push(currentNode.right);
        }
    }

    return list;
}

// Recursive:
const breadthFirstSearchR = function (queue, list) {
    if (!queue.length) {
        return list;
    }
    const currentNode = queue.shift();
    list.push(currentNode.value);

    if(currentNode.left) {
        queue.push(currentNode.left);
    }
    if(currentNode.right) {
        queue.push(currentNode.right);
    }

    return breadthFirstSearchR(queue, list);
}

File: searching_algos/BFS.py
# === BFS: Breadth First Search ===

# Strategy: Start with root node, then look left to right at the second level, then left to right on the third level (so on and so forth) until the target value is found. 

# Big O:
    # Time: O(n)

# IMPLEMENTATION:

def breadth_first_search(self):
    current_node = self.root
    result = []
    queue = []
    queue.append(current_node)

    while len(queue) > 0:
        current_node = queue.pop(0)
        result.append(current_node.value)
        if current_node.left:
            queue.append(current_node.left)
        if current_node.right:
            queue.append(current_node.right)
    
    return result

# Recursive
def breadth_first_search_R(self, queue, array):
    if not len(queue):
        return array
    current_node = queue.pop(0)
    array.append(current_node.value)

    if current_node.left:
        queue.append(current_node.left)
    if current_node.right:
        queue.append(current_node.right)

    return breadth_first_search_R(queue, array)
